src/main/scala/chiselverify/approximation/CompoundMetrics.scala

```diff
@@ -0,0 +1,520 @@
+package chiselverify.approximation
+
+import scala.math.{Fractional, Integral}
+
+import chiselverify.approximation.Metrics.MSE
+
+/** 
+  * This implementation assumes that all compound metrics are instantaneous.
+  * It also does not currently distinguish between absolute and relative
+  * metrics, per say.
+  * 
+  * Beware that the current interface supports samples only in either `BigInt`s 
+  * that are converted to `Double`s before any computations are carried out.
+  */
+object CompoundMetrics {
+
+  /** 
+    * Introduce some type abbreviations here for clarity.
+    */
+  type Signal = Iterable[BigInt]
+  type Image  = Iterable[Iterable[Iterable[BigInt]]]
+
+  /** 
+    * Explicit type conversion of an integral vector into a `BigInt` signal
+    * @param vctr the integral vector
+    * @return the same vector but converted into `BigInt`
+    */
+  def toSignal[T : Integral](vctr: Iterable[T]): Signal = {
+    vctr.map(elem => BigInt(implicitly[Integral[T]].toLong(elem)))
+  }
+
+  /** 
+    * Explicit type conversion of a fractional vector into a `BigInt` signal
+    * @param vctr the fractional vector
+    * @param scale the scale applied to elements in the signal (defaults to 1)
+    * @return the same matrix but converted into `BigInt` through scaling by `scale`
+    */
+  def toSignal[T : Fractional](vctr: Iterable[T], scale: Double = 1.0): Signal = {
+    val (min, max) = (Long.MinValue.toDouble, Long.MaxValue.toDouble)
+    vctr.map { elem =>
+      val prod = implicitly[Fractional[T]].toDouble(elem) * scale
+      assume(min <= prod && prod <= max)
+      BigInt(prod.longValue())
+    }
+  }
+
+  /** 
+    * Explicit type conversion of an integral matrix into a `BigInt` image
+    * @param mtrx the integral matrix
+    * @return the same matrix but converted into `BigInt`
+    */
+  def toImage[T : Integral](mtrx: Iterable[Iterable[Iterable[T]]]): Image = {
+    mtrx.map {
+      _.map {
+        _.map(elem => BigInt(implicitly[Integral[T]].toLong(elem)))
+      }
+    }
+  }
+
+  /** 
+    * Explicit type conversion of a fractional matrix into `BigInt` image
+    * @param mtrx the fractional matrix
+    * @param scale the scale applied to pixels in the image (defaults to 1)
+    * @return the same matrix but converted into `BigInt` through scaling by `scale`
+    */
+  def toImage[T : Fractional](mtrx: Iterable[Iterable[Iterable[T]]], scale: Double = 1.0): Image = {
+    val (min, max) = (Long.MinValue.toDouble, Long.MaxValue.toDouble)
+    mtrx.map {
+      _.map {
+        _.map { elem =>
+          val prod = implicitly[Fractional[T]].toDouble(elem) * scale
+          assume(min <= prod && prod <= max)
+          BigInt(prod.longValue())
+        }
+      }
+    }
```

Comment:
Can't this simply be a flatmap ? or do you necessarily want to maintain this triple depth ? 

Also why do you need the assumption here? Isn't that unsafe? Wouldn't an assertion be more appropriate ?

Reply:
I would like to keep the triple depth to stay compatible with the ImageMetrics below.

The assumption is hereby changed to an assertion. The distinction between when to use either isn't super clear to me. Do you have a good reference?

Reply:
So assumptions will be predicates that are considered a true in all subsequent assertions and requirements. This means that assumptions don't actually check anything they simply evaluate future assertions with added information that a given predicate can simply be *assumed* to be true. They are generally useful in a more modular setting where they can be used on already checked postconditions of a module you are instantiating in order to improve/speedup verification. They are generally risky and dangerous to use in most other situations.

Reply:
I see. I will refrain from including them in the middle of code in the future then. Thanks!